docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+# ADR 011: kURL Add-on Operator
+
+## Context
+
+kURL today is a set of bash scripts that configure and install a Kubernetes cluster in addition to a set of add-on components. The components are installed in sequence by applying highly customized Kubernetes YAML resources. The current architecture makes kURL hard to develop, maintain, and support.
+
+As an amalgam of bash scripts, the code base suffers from low cohesion and high complexity which results in the code being difficult to read and reason, and difficult to test. The installation is applied sequentially and each component is highly dependent on each other with no scopes to limit access and enforce separation. The project is architecturally a monolith without any functional separation of components and concerns.
+
+The installation of all components is done procedurally in the script. When there is a failure, it is difficult to understand what components have failed and how that affects the installation process as a whole.
+
+The components are installed in a proprietary fashion by applying highly customized Kubernetes YAML resources. They are hard to customize as the spec is proprietary and hard to extend because each customization must be passed through configuration to the end-user. Changes to this spec must be documented and are less discoverable. This method is not supported by the upstream which makes it difficult to pin addon installation issues on the upstream project and difficult to request support from the upstream.
+
+
+## Decision
+
+We will develop a Kubernetes [Operator](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) in _Go_ that will install, configure and manage add-on applications in the cluster. The operator will extend the Kubernetes API through Custom Resource Definitions (CRD) and will allow users to declaritively manage add-ons with a Custom Resource (CR). 
+
+## Solution
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
```

Comment:
I think this one we do not know yet right.

Reply:
I don't think that the ADR mentions backward compatibility. The phrasing seems correct: as long as the kURL installation has enabled only Kubernetes, Containerd and Flannel it should work. On that regard a kURL with Kubernetes, Containerd and Flannel is not different from k0s, kind or minikube.

Reply:
I got this objective from @emosbaugh add-on operator proposal [doc](https://docs.google.com/document/d/1ZnwgJxRZwxQKwEWCl9UQx7NRILk4hd28dwItPBuZjLU/edit#heading=h.isylz7bayrpk). I _think_ the point here is that it should work on a kURL Kubernetes minimal distro.